pvlib/singlediode_methods.py

```diff
@@ -0,0 +1,344 @@
+"""
+Faster ways to calculate single diode model currents and voltages using
+methods from J.W. Bishop (Solar Cells, 1988).
+"""
+
+from functools import partial
+import numpy as np
+try:
+    from scipy.optimize import brentq
+except ImportError:
+    brentq = NotImplemented
+# FIXME: change this to newton when scipy-1.2 is released
+try:
+    from scipy.optimize import _array_newton
+except ImportError:
+    from pvlib import tools
+    from pvlib.tools import _array_newton
+# rename newton and set keyword arguments
+newton = partial(_array_newton, tol=1e-6, maxiter=100, fprime2=None)
+
+
+def estimate_voc(photocurrent, saturation_current, nNsVth):
+    """
+    Rough estimate of open circuit voltage useful for bounding searches for
+    ``i`` of ``v`` when using :func:`~pvlib.pvsystem.singlediode`.
+
+    Parameters
+    ----------
+    photocurrent : numeric
+        photo-generated current [A]
+    saturation_current : numeric
+        diode one reverse saturation current [A]
+    nNsVth : numeric
+        product of thermal voltage ``Vth`` [V], diode ideality factor ``n``,
+        and number of series cells ``Ns``
+
+    Returns
+    -------
+    numeric
+        rough estimate of open circuit voltage [V]
+
+    Calculating the open circuit voltage, :math:`V_{oc}`, of an ideal device
+    with infinite shunt resistance, :math:`R_{sh} \\to \\infty`, and zero series
+    resistance, :math:`R_s = 0`, yields the following equation [1]. As an
+    estimate of :math:`V_{oc}` it is useful as an upper bound for the bisection
+    method.
+
+    .. math::
+
+        V_{oc, est}=n Ns V_{th} \\log \\left( \\frac{I_L}{I_0} + 1 \\right)
+
+    [1] http://www.pveducation.org/pvcdrom/open-circuit-voltage
+    """
+
+    return nNsVth * np.log(np.asarray(photocurrent) / saturation_current + 1.0)
+
+
+def bishop88(diode_voltage, photocurrent, saturation_current, resistance_series,
+             resistance_shunt, nNsVth, gradients=False):
+    """
+    Explicit calculation single-diode-model (SDM) currents and voltages using
+    diode junction voltages [1].
+
+    [1] "Computer simulation of the effects of electrical mismatches in
+    photovoltaic cell interconnection circuits" JW Bishop, Solar Cell (1988)
+    https://doi.org/10.1016/0379-6787(88)90059-2
+
+    Parameters
+    ----------
+    diode_voltage : numeric
+        diode voltages [V]
+    photocurrent : numeric
+        photo-generated current [A]
+    saturation_current : numeric
+        diode one reverse saturation current [A]
+    resistance_series : numeric
+        series resistance [ohms]
+    resistance_shunt: numeric
+        shunt resistance [ohms]
+    nNsVth : numeric
+        product of thermal voltage ``Vth`` [V], diode ideality factor ``n``,
+        and number of series cells ``Ns``
+    gradients : bool
+        default returns only i, v, and p, returns gradients if true
```

Comment:
`False returns only I, v, and p. True also returns gradients`

Reply:
✔️